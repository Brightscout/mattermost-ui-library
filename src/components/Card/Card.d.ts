import React from 'react';

export interface CardProps {

    /**
     * Content of the card
     */
    children?: React.ReactNode;

    /**
	 * To override or extend the styles applied to the component
	 */
    className?: string;
}
