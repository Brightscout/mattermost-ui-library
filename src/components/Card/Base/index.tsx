import React from 'react';

type BaseCardProps = {
    children: JSX.Element,
    className?: string;
}

const BaseCard = ({children, className = ''}: BaseCardProps) => (
    <div className={`wrapper padding-15 margin-bottom-15 channel-bg border-radius-8 ${className}`}>{children}</div>
);

export default BaseCard;
