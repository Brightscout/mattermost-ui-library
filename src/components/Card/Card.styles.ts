import styled from 'styled-components';

import Colors from '@Styles/colorsForJs.module.scss';

// Styles for card wrapper
export const CardWrapper = styled.div(() => {
    return {
        border: `1px solid ${Colors.centerChannel_16}`,
        borderRadius: '4px',
        background: Colors.centerChannelBg,
        boxShadow: Colors.elevation1,
        padding: 16,
    };
});
