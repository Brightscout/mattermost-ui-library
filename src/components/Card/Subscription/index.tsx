import React, {useMemo} from 'react';

// Components
import BaseCard from '@Components/Card/Base';
import Popover from '@Components/Popover';
import MenuButtons from '@Components/Buttons/MenuButtons';

// Constants
import {General as MMConstants} from 'mattermost-redux/constants';

type SubscriptionCardProps = {
    header: string | JSX.Element;
    label?: string;
    cardBody?: SubscriptionCardBody;
    description?: string;
    channel?: ChannelData | null;
    onDelete: (e: React.MouseEvent<HTMLButtonElement>) => void;
    onEdit: (e: React.MouseEvent<HTMLButtonElement>) => void;
    editIcon?: JSX.Element;
    deleteIcon?: JSX.Element;
    className?: string;
}

const SubscriptionCard = ({
    header,
    label,
    cardBody,
    description,
    onDelete,
    onEdit,
    channel,
    editIcon,
    deleteIcon,
    className = '',
}: SubscriptionCardProps) => {
    const buttonMenuPopover = useMemo(() => (
        <Popover
            popoverBody={
                <MenuButtons
                    buttons={[
                        {
                            icon: editIcon ?? 'fa fa-pencil-square-o',
                            text: 'Edit',
                            onClick: onEdit,
                        },
                        {
                            icon: deleteIcon ?? 'fa fa-trash-o',
                            text: 'Delete',
                            onClick: onDelete,
                        },
                    ]}
                />
            }
            className='subscription-card__popover'
        >
            <button className='style--none subscription-card__menu-btn'>
                <i className='fa fa-ellipsis-v'/>
            </button>
        </Popover>
    ), [onEdit, onDelete]);

    return (
        <BaseCard className={`subscription-card margin-bottom-15 ${className}`}>
            <>
                <div className='margin-bottom-10 d-flex justify-content-between'>
                    <p className='margin-0 wt-500'>{header}</p>
                    {buttonMenuPopover}
                </div>
                {channel && (
                    <div className='subscription-card__channel margin-bottom-12 padding-h-15 d-flex align-items-center'>
                        <i className={`channel-icon margin-right-5 ${channel.type === MMConstants.PRIVATE_CHANNEL ? 'icon icon-lock-outline' : 'icon icon-globe'}`}/>
                        <span className='subscription-card__channel-text font-14 text-ellipsis'>{channel.display_name}</span>
                    </div>
                )}
                {label && <div className='subscription-card__label padding-v-2 padding-h-5 text-capitalize font-size-10 wt-600 margin-bottom-10 channel-text d-inline-block border-radius-4'>{label}</div>}
                {(cardBody || description) && (
                    <ul className='subscription-card__body padding-0 margin-0 font-size-14'>
                        {cardBody?.list?.map((listItem: string | JSX.Element, index: number) => (
                            <li
                                key={index}
                                className='subscription-card__body-item margin-bottom-5 margin-left-15 subscription-card__body-item--list font-12'
                            >
                                {listItem}
                            </li>
                        ))}
                        {cardBody?.labelValuePairs?.map((bodyItem, index: number) => (
                            <li
                                key={bodyItem.label}
                                className={`text-ellipsis subscription-card__body-item margin-bottom-10 ${cardBody?.list?.length && !index && 'margin-top-15'}`}
                            >
                                <span className='subscription-card__body-header wt-600 margin-right-10'>{bodyItem.label + ':'}</span>
                                <span className='subscription-card__body-text'>{bodyItem.value}</span>
                            </li>
                        ))}
                        {description && (
                            <li
                                className='subscription-card__body-item'
                            >
                                {description}
                            </li>
                        )}
                    </ul>
                )}
            </>
        </BaseCard>
    );
};

export default SubscriptionCard;
