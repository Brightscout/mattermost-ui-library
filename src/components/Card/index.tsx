
import React from 'react';

// Icons
import {DeleteIcon, EditIcon} from '../Icons';

type CardProps = {
    cardHeader: string;
    cardSubHeader?: string;
    cardBody: CardBody[];
    onDelete?: (e: React.MouseEvent<HTMLButtonElement>) => void;
    onEdit?: (e: React.MouseEvent<HTMLButtonElement>) => void;
    editIcon?: JSX.Element;
    deleteIcon?: JSX.Element;
    className?: string;
}

const Card = ({
    cardHeader,
    cardSubHeader,
    cardBody,
    onDelete,
    onEdit,
    editIcon,
    deleteIcon,
    className = '',
}: CardProps) => (
    <div className={`details-card border-radius-8 margin-bottom-15 ${className}`}>
        <div className='details-card__header padding-v-12 padding-h-25 sidebar-bg d-flex align-items-center justify-content-between'>
            <h3 className='details-card__header-text font-16 margin-0 wt-500 text-ellipsis'>
                {cardHeader}
                {cardSubHeader && <span className='details-card__sub-header margin-left-5'>{cardSubHeader}</span>}
            </h3>
            <div>
                {onEdit && (
                    <button
                        onClick={onEdit}
                        className='details-card__action-btn padding-0 margin-right-15 plugin-btn'
                    >
                        {editIcon ?? <EditIcon />}
                    </button>
                )}
                {onDelete && (
                    <button
                        onClick={onDelete}
                        className='details-card__action-btn padding-0 plugin-btn'
                    >
                        {deleteIcon ?? <DeleteIcon />}
                    </button>
                )}
            </div>
        </div>
        <div className='card-body padding-v-5'>
            {cardBody.map((body) => (
                <div
                    key={body.sectionHeading as string}
                    className='card-body__section margin-v-12 margin-h-25'
                >
                    <h3 className='card-body__section-heading wt-600 font-14 margin-bottom-5 text-ellipsis'>{body.sectionHeading}</h3>
                    <ul className='card-body__list padding-0 margin-bottom-10'>
                        {body.data.map((listItem) => (
                            <li
                                key={listItem.label as string}
                                className='body-item'
                            >
                                <p className='body-item__text font-14 text-ellipsis'>
                                    <span className='body-item__icon margin-right-10'>{listItem.icon}</span>
                                    <span>{listItem.label}</span>
                                </p>
                            </li>
                        ))}
                    </ul>
                </div>
            ))}
        </div>
    </div>
);

export default Card;
