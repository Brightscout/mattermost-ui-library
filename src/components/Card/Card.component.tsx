import React from 'react';

import {CardProps} from './Card';
import {CardWrapper} from './Card.styles';

/**
 * Card Component
 *
 * @example Correct usage
 *
 * ```ts
 * <Card> Content </Card>
 * ```
 */
export const Card = (props: CardProps) => {
    const {children, className = ''} = props;

    return <CardWrapper className={`mm-card ${className}`}>{children}</CardWrapper>;
};
