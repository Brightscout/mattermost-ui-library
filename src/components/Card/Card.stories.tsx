import React from 'react';
import {Story, Meta} from '@storybook/react';

import {Card} from './Card.component';
import {CardProps} from './Card';

export default {
    title: 'Component/Card',
    component: Card,
} as Meta<typeof Card>;

const CardTemplate: Story<CardProps> = (args) => <Card {...args}/>;

// Default
export const Default = CardTemplate.bind({});
Default.args = {
    children: 'Passed in children',
};
