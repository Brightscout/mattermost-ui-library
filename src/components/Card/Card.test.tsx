import React from 'react';
import {shallow, ShallowWrapper} from 'enzyme';

import Card from '.';

// Mock functions
const editHandler = jest.fn();
const deleteHandler = jest.fn();

// Mock data for cardBody props
const cardBody = [{
    sectionHeading: 'Section Heading',
    data: [{
        icon: <i className='fa fa-edit' />,
        label: 'Label 1',
    }, {
        icon: <i className='fa fa-link' />,
        label: 'Label 2',
    }],
}];

// Mock Icons
const EditIcon = <i className='fa fa-edit' />;
const DeleteIcon = <i className='fa fa-delete' />;

// Mock Props
const cardProps = {
    cardHeader: 'Card Header',
    cardSubHeader: 'Card sub header',
    cardBody,
    onEdit: editHandler,
    onDelete: deleteHandler,
    editIcon: EditIcon,
    deleteIcon: DeleteIcon,
    className: 'card-custom-class',
}


describe('Card', () => {
    let component: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>

    beforeEach(() => {
        component = shallow(<Card {...cardProps} />);
    });

    it('Should render correctly', () => {
        expect(component).toMatchSnapshot();
    });

    it('Should apply the passed className prop', () => {
        expect(component.hasClass(cardProps.className)).toBeTruthy();
    })

    it('Should render the header correctly', () => {
        expect(component.find('.details-card__header-text').text().includes(cardProps.cardHeader)).toBeTruthy();
    })

    it('Should render the sub-header correctly', () => {
        expect(component.find('.details-card__sub-header').text().includes(cardProps.cardSubHeader)).toBeTruthy();
    })

    it('Should render the card body correctly', () => {
        expect(component.find('.card-body__section')).toHaveLength(cardBody.length);
    })

    it('Should render the passed edit and delete Icons', () => {
        expect(component.containsMatchingElement(EditIcon)).toBeTruthy();
        expect(component.containsMatchingElement(DeleteIcon)).toBeTruthy();
    });
})
