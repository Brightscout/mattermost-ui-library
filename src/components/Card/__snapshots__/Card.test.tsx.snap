// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Should render correctly 1`] = `
<div
  className="details-card border-radius-8 margin-bottom-15 card-custom-class"
>
  <div
    className="details-card__header padding-v-12 padding-h-25 sidebar-bg d-flex align-items-center justify-content-between"
  >
    <h3
      className="details-card__header-text font-size-16 margin-0 wt-500 text-ellipsis"
    >
      Card Header
      <span
        className="details-card__sub-header margin-left-5"
      >
        Card sub header
      </span>
    </h3>
    <div>
      <button
        className="details-card__action-btn padding-0 margin-right-15"
        onClick={[MockFunction]}
      >
        <i
          className="fa fa-edit"
        />
      </button>
      <button
        className="details-card__action-btn padding-0"
        onClick={[MockFunction]}
      >
        <i
          className="fa fa-delete"
        />
      </button>
    </div>
  </div>
  <div
    className="card-body padding-v-5"
  >
    <div
      className="card-body__section margin-v-12 margin-h-25"
      key="Section Heading"
    >
      <h3
        className="card-body__section-heading wt-600 font-size-14 margin-bottom-5 text-ellipsis"
      >
        Section Heading
      </h3>
      <ul
        className="card-body__list padding-0 margin-bottom-10"
      >
        <li
          className="body-item"
          key="Label 1"
        >
          <p
            className="body-item__text font-size-14 text-ellipsis"
          >
            <span
              className="body-item__icon margin-right-10"
            >
              <i
                className="fa fa-edit"
              />
            </span>
            <span>
              Label 1
            </span>
          </p>
        </li>
        <li
          className="body-item"
          key="Label 2"
        >
          <p
            className="body-item__text font-size-14 text-ellipsis"
          >
            <span
              className="body-item__icon margin-right-10"
            >
              <i
                className="fa fa-link"
              />
            </span>
            <span>
              Label 2
            </span>
          </p>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
